proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,260 @@
+
+Cons:
+* Extremely heavy process that will make us afraid/refuse to make improvements to WAL, because it's too much work. It might fails our goal of `Balancing development velocity with user data stability risks`
+  * One mitigation would be an LTS retroactive strategy e.g. LTS 3.1 only supports WALv1, 3.3 adds WALv2, we do 3.1.1 with WALv2 too, 3.4 can now switch to WALv2, 4.0 can remove WALv1 support. It gives us more flexibility, but it's not very realistic to do patch release of LTS with risky feature like a new WAL.
+* We literally have no formal process for LTS versions and we don't do them regularly.
+
+### Add a CLI tool that rewrites WAL to a specific version
+
+We could add a `promtool` command that rewrites WAL segments to a given version.
+
+Users can then migrate their WAL with a single command either as an init process before reverting or manually on a disk.
+
+This should take ~minutes even for larger WAL files.
+
+Pros:
+* Less need for a two-fold strategy?
+
+Cons:
+* We need to write migration code and ensure it's efficient enough for bigger data
+* A bit painful to use on scale and remotely (e.g. on Kubernetes)
+
+### Rewrite WAL before/during replay
+
+Instead of supporting multiple directories for WAL for various versions, we could rewrite WAL on the start.
+
+Pros:
+* Simpler implementation
+
+Cons:
+* We already suffer from replay problems, so I propose an eventual rewrite.
+* Rewrite is more risking than read-only WAL (of previous version)
+
+### Don't version WAL, don't introduce a flag
+
+Pro:
+* One issue with a flag is that we lose "automatic" rollout for some users e.g. what if we will have another v3 WAL version, but users sets an explicit `--storage.tsdb.stateful.write-wal-version=v3` option? For those users they will need to make a manual choice and there might be not aware of benefits of a new WAL or do any work to test new WAL for their setup. With raw two-fold strategy format is forced, but also transparent (if without issues).
+
+Cons:
+* Takes more time for features to be available to users
+* Demotivating for format changes (long feedback loop)
+* Harder to communicate what exactly changed in each Prometheus version or even implement backward compatibility?
+
+### Record-based or segment based WAL versioning
+
+Given we usually change WAL by changing its records, the WAL version could be simply [max number of types](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/record/record.go#L54) we write to WAL.
+Alternatively, it could be per segment e.g. introduce a special version record type that is only in the front of the segment file.
+
+Pros:
+* Simpler to implement now.
+* This would allow mixing Y and Y+1 segments/records in the same WAL, we would not need a new directory.
+
+Cons:
+* Won't work for bigger changes e.g.
+  * changes that merge records 
+  * sharding?
+
+### Maintain two WALs (well four, with WBL)
```

Comment:
why do we have separate WAL and WBL ?

Reply:
Because WBL is a separate WAL on disk

Reply:
That's a re-statement of the fact.  I am asking what the reason behind it is.

Reply:
When implementing out-of-order, the WAL had all the out-of-order samples in them (both the ingested and uningested out-of-order samples). So we needed a way to know exactly what out-of-order samples were ingested, so that we can restore only those samples. We could only know if a sample went into a chunk at the time of writing to the chunk (i.e. after writing to WAL), so we introduced WBL to log the out-of-order samples ingested, since it was logged after samples went into the chunk. Data in WBL only goes into the out-of-order head.

Reply:
This could probably be merged together by writing separate OOO records in WAL. But the replay needs to be changed carefully so that things don't break.